More declarative timeout syntax for manual alarm control panel

bonzini opened the pull request:
## Description:

This series adds new settings to the manual alarm control panel, so that the state machine is able to express many common use cases in an entirely declarative manner, without relying on "for" and "condition" checks. For more information, see the doc pull request. The changes are all included in the MQTT version as well.

In addition, it adds the possibility to compute the code with a jinja template, thus enabling integration with the OTP sensor.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4039

## Example entry for `configuration.yaml` (if applicable):
```yaml
alarm_control_panel:
  - platform: manual
    code: 1234
    pending_time: 30
    delay_time: 20
    trigger_time: 4
    disarmed:
      trigger_time: 0
    armed_home:
      pending_time: 0
      delay_time: 0
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

bonzini committed:
alarm/manual, alarm/manual_mqtt: cleanup pending_time

Change it to cv.time_period, remove it from ManualAlarm constructor
since it is always filled in for each individual alarm state.

bonzini committed:
alarm/manual, alarm/manual_mqtt: fix doc comment

bonzini committed:
alarm/manual, alarm/manual_mqtt: add template-based code

This lets you specify a code only when disarming, use OTP sensors
to generate codes, and much more.  The only limit is your imagination!

bonzini committed:
alarm/manual, alarm/manual_mqtt: add delay_time setting

The delay_time argument is complementary to pending_time.  Instead of
causing the alarm to become PENDING when switching to a state, it
causes the alarm to become PENDING when switching *from* a state to the
TRIGGERED state.

For example, you could set a delay_time to 30 if opening the garage door
will take 30 seconds, during which the dog could walk in front of the
door and trigger a presence alarm.

The delay time can also be set per state; for example, at night doors
are supposed to be closed, and a presence alarm would necessarily be
triggered "by the inside"; so "armed_night" would have zero delay_time.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
alarm/manual, alarm/manual_mqtt: add pre_pending_state attribute

Currently it's not clear what prevents a switch from a state "to
itself" while the transition is pending.  Extend the _pre_trigger_state
infrastructure to avoid that case, and expose it as a new attribute for
users and tests.  It will also come in handy when implementing per-state
delay and trigger times.

Tests are added in a later patch (currently, tests do not cover
post_pending_state either).

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
alarm/manual_mqtt: sync different test with alarm/manual

Make sure that both tests try the same scenarios, so that the same
changes can be applied to both.

homeassistant commented:
Hi @bonzini, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10738).

Once you do that we will be able to review and accept this pull request.

Thanks!

bonzini commented:
For a better idea of where I am going with these changes, see [here](https://github.com/bonzini/home-assistant.github.io/blob/alarm-group/source/_cookbook/automation_alarm.markdown) which would be my next pull request.

The "group" component used in the linked cookbook recipe allows for an even easier and more declarative alarm configuration. Code-wise, it relies on the pre_pending_state attribute introduced by this pull request in order to compute the state of the group panel.

MartinHjelmare commented:
After reading your documentation PR, I think this makes sense and is a good addition. I've skimmed the code, but it's a quite many lines so I'll go through it more later if nobody else claims this.

bonzini commented:
Thanks! Of course most of it is tests, the logic changes are relatively small and the backwards-compatibility should be almost self-evident.

For the future, do you prefer to change tests in separate commits or in the same?

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual.py:
```diff
@@ -16,53 +16,85 @@
     vol.Optional(CONF_PENDING_TIME, default=DEFAULT_PENDING_TIME):
         vol.All(vol.Coerce(int), vol.Range(min=0)),
     vol.Optional(CONF_TRIGGER_TIME, default=DEFAULT_TRIGGER_TIME):
-        vol.All(vol.Coerce(int), vol.Range(min=1)),
```
Do you know why `CONF_TRIGGER_TIME` it was set to min 1 before?

MartinHjelmare replied:
Maybe add this as a comment here:
`# A trigger_time of zero disables the alarm_trigger service.`

bonzini replied:
I don't, but `CONF_PENDING_TIME` also started as min=1 and then was changed to min=0. It seems to me that leaving min=1 for `CONF_TRIGGER_TIME` was an oversight when adding those `if self._trigger_time:` checks.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual.py:
```diff
@@ -86,23 +120,36 @@ class ManualAlarm(alarm.AlarmControlPanel):
+            self._code = code_template
+            self._code.hass = hass
+        else:
+            self._code = str(code) if code else None
```
The code is already validated as a string if set in the config. If we want to check for empty string, I think we should do that already in the config validation. Could you add a validator that raises vol.Invalid for empty string? You should be able to use `vol.IsTrue()`.

bonzini replied:
I think the purpose here is to make `code = ''` equivalent to no code at all. I'll change it to `self._code = code or None`.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual.py:
```diff
@@ -86,23 +120,36 @@ class ManualAlarm(alarm.AlarmControlPanel):
+        self._delay_time_by_state = {}
+        self._trigger_time_by_state = {}
+        for state in SUPPORTED_PRETRIGGER_STATES:
+            self._delay_time_by_state[state] = datetime.timedelta(
```
Use the `time_period_seconds` validator from config_validation.py during config validation.

bonzini replied:
This was just copied from pending time handling, so I'll clean that one up as well.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual.py:
```diff
@@ -86,23 +120,36 @@ class ManualAlarm(alarm.AlarmControlPanel):
+        for state in SUPPORTED_PRETRIGGER_STATES:
+            self._delay_time_by_state[state] = datetime.timedelta(
+                seconds=config[state][CONF_DELAY_TIME])
+            self._trigger_time_by_state[state] = datetime.timedelta(
```
Same as above.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual.py:
```diff
@@ -86,23 +120,36 @@ class ManualAlarm(alarm.AlarmControlPanel):
+                seconds=config[state][CONF_TRIGGER_TIME])
         self._pending_time_by_state = {}
         for state in SUPPORTED_PENDING_STATES:
             self._pending_time_by_state[state] = datetime.timedelta(
```
Same as above.

MartinHjelmare reviewed (commented):
I've only looked at the manual platform so far and no tests. You need to rebase on dev too, and solve the conflicts.

bonzini committed:
alarm/manual, alarm/manual_mqtt: cleanup trigger_time

Change it to cv.time_period; in order to use cv.positive_timedelta,
fix up the code to allow specifying a zero trigger time.

A zero trigger time disables triggering the alarm.  The code is more
or less already there, but was not active because the schema requests a
minimum of one second for the trigger_time.  This is not very useful as
is, but it will be after the next patch to make the trigger time
customizable per-state.

bonzini committed:
alarm/manual, alarm/manual_mqtt: add per-state trigger_time

During a pending state, the trigger_time is the one of the previous state.
This lets the user implement a "true" disarmed state, without having to
test it in the conditions for the alarm control panel's state.  This
simplifies the automation if you are not interested in multiple states.
For example you can just add

  disarmed:
    trigger_time: 0

in the manual alarm and omit the

  condition:
    - condition: state
      entity_id: alarm_control_panel.ha_alarm
      state: armed_away

in all the actions.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual.py:
```diff
@@ -186,33 +240,45 @@ def alarm_arm_custom_bypass(self, code=None):
         """Send alarm trigger command. No code needed."""
-        self._pre_trigger_state = self._state
+        if not self._trigger_time_by_state[self._active_state]:
+            return
```
Add a comment here why, ie "A trigger_time of zero disables the alarm_trigger service."

MartinHjelmare commented on homeassistant/components/alarm_control_panel/manual_mqtt.py:
```diff
@@ -216,33 +270,45 @@ def alarm_arm_night(self, code=None):
         """Send alarm trigger command. No code needed."""
-        self._pre_trigger_state = self._state
+        if not self._trigger_time_by_state[self._active_state]:
+            return
```
Same as above.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -559,6 +788,211 @@ def test_disarm_during_trigger_with_invalid_code(self):
+        assert state.state == STATE_ALARM_TRIGGERED
+
+    def test_trigger_with_pending_and_specific_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -674,45 +1108,145 @@ def test_trigger_with_specific_pending(self):
 
-        future = dt_util.utcnow() + timedelta(seconds=2)
+    def test_arm_away_after_disabled_disarmed(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -559,6 +788,211 @@ def test_disarm_during_trigger_with_invalid_code(self):
+        assert state.state == STATE_ALARM_TRIGGERED
+
+    def test_trigger_with_pending_and_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -559,6 +788,211 @@ def test_disarm_during_trigger_with_invalid_code(self):
+        assert state.state == STATE_ALARM_TRIGGERED
+
+    def test_trigger_with_specific_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -559,6 +788,211 @@ def test_disarm_during_trigger_with_invalid_code(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_trigger_with_unused_specific_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -425,6 +553,107 @@ def test_trigger_with_disarm_after_trigger(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_trigger_with_zero_specific_trigger_time(self):
+        """Test disarm after trigger."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -345,6 +380,99 @@ def test_trigger_no_pending(self):
+                         self.hass.states.get(entity_id).state)
+
+    def test_trigger_zero_trigger_time_with_pending(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -345,6 +380,99 @@ def test_trigger_no_pending(self):
+        self.assertEqual(STATE_ALARM_TRIGGERED, state.state)
+
+    def test_trigger_zero_trigger_time(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -345,6 +380,99 @@ def test_trigger_no_pending(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_trigger_with_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual_mqtt.py:
```diff
@@ -162,6 +162,34 @@ def test_arm_away_no_pending(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_arm_home_with_template_code(self):
+        """Attempt to arm home without a valid code."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -795,3 +1246,75 @@ def test_armed_custom_bypass_with_specific_pending(self):
                          self.hass.states.get(entity_id).state)
+
+    def test_arm_away_after_disabled_disarmed(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -518,6 +835,101 @@ def test_trigger_with_disarm_after_trigger(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_trigger_with_zero_specific_trigger_time(self):
+        """Test disarm after trigger."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -355,6 +475,203 @@ def test_trigger_with_pending(self):
+        assert state.state == STATE_ALARM_TRIGGERED
+
+    def test_trigger_with_pending_and_specific_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -355,6 +475,203 @@ def test_trigger_with_pending(self):
+        assert state.state == STATE_ALARM_TRIGGERED
+
+    def test_trigger_with_pending_and_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -355,6 +475,203 @@ def test_trigger_with_pending(self):
+        assert state.state == STATE_ALARM_TRIGGERED
+
+    def test_trigger_with_specific_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -355,6 +475,203 @@ def test_trigger_with_pending(self):
         assert state.state == STATE_ALARM_DISARMED
 
+    def test_trigger_with_unused_specific_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -311,6 +344,93 @@ def test_trigger_no_pending(self):
+                         self.hass.states.get(entity_id).state)
+
+    def test_trigger_zero_trigger_time_with_pending(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -311,6 +344,93 @@ def test_trigger_no_pending(self):
+        self.assertEqual(STATE_ALARM_TRIGGERED, state.state)
+
+    def test_trigger_zero_trigger_time(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -311,6 +344,93 @@ def test_trigger_no_pending(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_trigger_with_delay(self):
+        """Test arm home method."""
```
Stale docstring.

MartinHjelmare commented on tests/components/alarm_control_panel/test_manual.py:
```diff
@@ -140,6 +140,32 @@ def test_arm_away_no_pending(self):
                          self.hass.states.get(entity_id).state)
 
+    def test_arm_home_with_template_code(self):
+        """Attempt to arm home without a valid code."""
```
Stale docstring.

MartinHjelmare reviewed (changes requested):
Add two comments as clarification and update docstrings for tests.

We should also consider labeling this PR as breaking, since the test `test_trigger_with_specific_pending` has changed.

bonzini commented:
I'll do the updates, thanks! I don't think the pull request should be _breaking_, because the changes to the tests are just widening the block where a mock is used (to avoid that time goes "backwards"). There is no behavior change when measuring real time.

MartinHjelmare reviewed (approved):
Ok, looks good.

For the future please don't squash the commits during the review process. That makes it very hard to follow the changes. I've done some sampling of the code and it looks good, but I won't go over everything again.
